include/util/cell_storage.hpp

```diff
@@ -0,0 +1,314 @@
+                std::find(destination_boundary, destination_boundary + num_destination_nodes, node);
+            BOOST_ASSERT(iter != destination_boundary + num_destination_nodes);
+            return iter - destination_boundary;
+        }
+
+      public:
+        auto GetOutWeight(NodeID node) const
+        {
+            auto row = GetRow(node);
+            auto begin = weights + num_destination_nodes * row;
+            auto end = begin + num_destination_nodes;
+            return boost::make_iterator_range(begin, end);
+        }
+
+        auto GetInWeight(NodeID node) const
+        {
+            auto column = GetColumn(node);
+            auto begin = ColumnIterator{weights + column, num_destination_nodes};
+            auto end = ColumnIterator{weights + column + num_source_nodes * num_destination_nodes,
+                                      num_destination_nodes};
+            return boost::make_iterator_range(begin, end);
+        }
+
+        auto GetSourceNodes() const
+        {
+            return boost::make_iterator_range(source_boundary, source_boundary + num_source_nodes);
+        }
+
+        auto GetDestinationNodes() const
+        {
+            return boost::make_iterator_range(destination_boundary,
+                                              destination_boundary + num_destination_nodes);
+        }
+
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
+            : num_source_nodes{data.num_source_nodes},
+              num_destination_nodes{data.num_destination_nodes},
+              weights{all_weight + data.weight_offset},
+              source_boundary{all_sources + data.source_boundary_offset},
+              destination_boundary{all_destinations + data.destination_boundary_offset}
+        {
+            BOOST_ASSERT(all_weight != nullptr);
+            BOOST_ASSERT(all_sources != nullptr);
+            BOOST_ASSERT(all_destinations != nullptr);
+        }
+    };
+
+    std::size_t LevelIDToIndex(LevelID level) const { return level - 1; }
+
+  public:
+    using Cell = CellImpl<EdgeWeight>;
+    using ConstCell = CellImpl<const EdgeWeight>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
+    {
+        // pre-allocate storge for CellData so we can have random access to it by cell id
+        unsigned number_of_cells = 0;
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            level_to_cell_offset.push_back(number_of_cells);
+            number_of_cells += partition.GetNumberOfCells(level);
+        }
+        level_to_cell_offset.push_back(number_of_cells);
+        cells.resize(number_of_cells);
+
+        std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+        std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
```

Comment:
What about making this proper types instead of using `.first` and `.second` below

    struct LevelNode { CellID level; NodeID node; };

    std::vector<LevelNode> sources, sinks;

Then we can say `.level` and `.node` instead.